src/library/scala/collection/Iterable.scala

```diff
@@ -549,21 +549,22 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
     *
     */
   def groupBy[K](f: A => K): immutable.Map[K, C] = {
-    val m = mutable.Map.empty[K, Builder[A, C]]
+    val m = mutable.HashMap.empty[K, Any /*C | Builder[A, C] */]
     val it = iterator
-    while (it.hasNext) {
-      val elem = it.next()
-      val key = f(elem)
-      val bldr = m.getOrElseUpdate(key, newSpecificBuilder)
-      bldr += elem
-    }
-    var result = immutable.HashMap.empty[K, C]
-    val mapIt = m.iterator
-    while (mapIt.hasNext) {
-      val (k, v) = mapIt.next()
-      result = result.updated(k, v.result())
+    if (it.isEmpty) immutable.Map.empty
+    else {
+      while (it.hasNext) {
+        val elem = it.next()
+        val key  = f(elem)
+        val bldr = m.getOrElseUpdate(key, newSpecificBuilder).asInstanceOf[Builder[A, C]]
+        bldr += elem
+      }
+      m.mapValuesInPlace((_, v) => v.asInstanceOf[Builder[A, C]].result())
```

Comment:
couldn't we combine these into a new Map function?
this could 

- avoid the casts as `m = mutable.HashMap.empty[K, Builder[A, C]]`
- dual iteration as we traverse, build and create as the same time
- tuple creation from mapValuesInPlace
- MapBuilder creation

```
Map.fromMutableTransformingValues(m, (_, v) => v.result())

in Map
private[scala] // for Mima :-(
def  fromMutableTransformingValues[A, B, C](m: mutable.Map[A,B], transform: Function2[A, B, C]): Map[A, C] {
…//similar to the customised addAdd in builder
}
```

Reply:
`mapValuesInPlace` doesn't create `Tuples` here, it is overriden in `mutable.HashMap`.

```
private[mutable] def mapValuesInPlaceImpl(f: (K, V) => V): this.type = {
    val len = table.length
    var i = 0
    while (i < len) {
      var n = table(i)
      while (n ne null) {
        n.value = f(n.key, n.value)
        n = n.next
      }
      i += 1
    }
    this
  }
```

It isn't worth avoiding the `HashMapBuilder` allocation IMO. I've already special cased the `isEmpty` case to make that allocation free.

Avoiding the dual traversal would be nice, but the traversals are likely to to come from cache for non-enourmous maps.

avoidnt